Sodium Encryption

pmclain opened the pull request:
### Description
Moves all new encryption to the sodium library. Supports decryption using legacy mcrypt ciphers. This PR ensures new values are (en-de)crypted with sodium and legacy values can be decrypted using mcrypt. 

A data patch updating configuration values with a `backend_model` of `\Magento\Config\Model\Config\Backend\Encrypted::class` and the contents of `sales_order_payment.cc_number_enc`. This mirrors the re-encryption behavior when change the key in `Magento\EncryptionKey\Model\ResourceModel\Key\Change`

### Fixed Issues (if relevant)
1. magento-engcom/php-7.2-support#124 
2. magento-engcom/php-7.2-support#128 - `\Magento\Framework\Encryption\Crypt` is deprecated and replaced with `\Magento\Framework\Encryption\Adapter\Mcrypt`. An exception is thrown when `Mcrypt::encrypt` is called.
3. magento-engcom/php-7.2-support#129

### Manual testing scenarios
1. Install any Magento version using mcrypt
2. Configure Braintree Payment Gateway in admin panel
3. Checkout using Braintree Payment
4. Expect success
5. Upgrade installation to libsodium
6. Checkout using Braintree Payment
7. Expect success - the existing mcrypt gateway credentials are decrypted to allow checkout. Values are **not** re-encrypted with sodium.
8. Update Braintree encrypted values in admin panel to trigger sodium encryption
9. Checkout using Braintree Payment
10. Expect success

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

pmclain committed:
Deprecate Crypt Class

pmclain committed:
Add Encryption Adapters

* Create `Magento\Framework\Encryption\Adapter\EncryptionAdapterInterface`
* Implement Mcrypt and Sodium Adapters

The Mcrypt adapter is implemented for decryption backwards compatability and
throws an exceptions if used for encryption.

pmclain committed:
Add Encryption Adapters

* Create `Magento\Framework\Encryption\Adapter\EncryptionAdapterInterface`
* Implement Mcrypt and Sodium Adapters

The Mcrypt adapter is implemented for decryption backwards compatability and
throws an exceptions if used for encryption.

pmclain committed:
Change Encyption to Sodium

* Add new cipher constant and update cosntant value for latest cipher
* All encrpytion is done using Sodium
* Decrpytion is done using the original cipher

pmclain committed:
Set Random Key Size to Match SODIUM_CRYPTO_AEAD_CHACHA20POLY1305_KEYBYTES

pmclain committed:
Require paragonie/sodium_compat

pmclain committed:
Clear sensitive information from memory

pmclain committed:
Revert clearing sensitive data

sodium_compat throws an exception when sodium_memzero is called because the PHP
implemenation cannot reliably zero buffers.

pmclain committed:
Fix Failing Integration Tests

* update regex validation for 3 parts and make closing = optional
* correct expected return type when calling getCrypt

pmclain committed:
Migrate encrypted values to sodium

buskamuza commented on app/code/Magento/EncryptionKey/Setup/Patch/Data/SodiumChachaPatch.php:
```diff
@@ -0,0 +1,129 @@
+
+/**
+ * Class SodiumChachaPatch
+ * @package Magento\EncryptionKey\Setup\Patch
```
Could you please replace auto-generated doc blocks with descriptions of the class/method?

buskamuza replied:
Also, please add type hinting for parameters and return types in the new methods.

buskamuza commented on lib/internal/Magento/Framework/Encryption/Adapter/Sodium.php:
```diff
@@ -0,0 +1,69 @@
+
+        return $this->keyVersion .
+            ':' . Encryptor::CIPHER_AEAD_CHACHA20POLY1305 .
+            ':' . base64_encode($nonce . $cipherText);
```
Encrypt-decrypt looks asymmetric, as knowledge of storage format is in two classes (during encryption we add key and cipher to the encrypted text in the adapter, and during decryption it happens in the Encryptor).
I'd better add `getKey()` and `getCipher()` methods (or similar) to the adapter interface, and use this information in the `Encryptor` to form the text for the storage.
This also help to design by contract. With current implementation interface does not guarantee that the encrypted data will be returned in the format "<key>:<cipher>:<ciphertext>". But with moving this logic to the `Encryptor`, we don't need these guarantees from the adapter.

buskamuza commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+<?php
```
Ideally, we want to reuse the code by refactoring the old class `Crypt` so that it calls methods of the new class.
It should be possible in this case, as the code is just the same, except for `encrypt()` method.

buskamuza commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+<?php
+
+namespace Magento\Framework\Encryption\Adapter;
+
```
Could you please add to the description an explanation of why we need this class?

buskamuza commented on lib/internal/Magento/Framework/Encryption/Encryptor.php:
```diff
@@ -279,6 +285,7 @@ public function encrypt($data)
      *
      * @param string $data
      * @return string
+     * @throws \Exception
```
Why do we get exception here? Is it Sodium lib throwing an exception in some cases?
If yes, please warp the call to sodium in `try-catch` and re-throw more specific exception.

> 5.7. Thrown exceptions SHOULD be as specific as possible. The top generic \Exception SHOULD NOT be thrown anywhere.
5.8. All direct communications with third-party libraries MUST be wrapped with a try/catch statement.

https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html#5-exceptions

buskamuza replied:
I know that we throw `\Exception` in `validate*()` methods. Those are from old times...

buskamuza commented on lib/internal/Magento/Framework/Encryption/Encryptor.php:
```diff
@@ -328,7 +335,7 @@ public function decrypt($data)
      *
      * @param string|null $key NULL value means usage of the default key specified on constructor
-     * @return \Magento\Framework\Encryption\Crypt
+     * @return EncryptionAdapterInterface
```
(?) Backward incompatible change

buskamuza commented on lib/internal/Magento/Framework/Encryption/Encryptor.php:
```diff
@@ -371,7 +379,8 @@ public function exportKeys()
      * @param int $cipherVersion
      * @param bool $initVector
-     * @return Crypt|null
+     * @return EncryptionAdapterInterface|null
```
(?) Backward incompatible change (related to another one)

buskamuza commented on app/code/Magento/EncryptionKey/Setup/Patch/Data/SodiumChachaPatch.php:
```diff
@@ -0,0 +1,129 @@
+
+    private function reEncryptCreditCardNumbers()
+    {
+        $table = $this->moduleDataSetup->getTable('sales_order_payment');
```
This data can be large.
I'd just leave the data encrypted, so it will be re-encrypted when read/written next time.
Additionally we could provide a separate tool or Magento CLI command to convert encrypted data. This can be run separately on already upgraded Magento instance. Both old and new data is usable for Magento application, so users can run the tool only if they really want to migrate to the newer algorithm (which should not be a common case now, as current algorithm still pretty good as I understand).

buskamuza commented on lib/internal/Magento/Framework/Encryption/Crypt.php:
```diff
@@ -10,6 +10,7 @@
  * Class encapsulates cryptographic algorithm
  *
  * @api
+ * @deprecated
```
Please add `@see` with the reference to the replacement.

buskamuza replied:
Please add the class to the static test that searches for obsolete code, so nobody tries to use it.

paliarush commented on app/code/Magento/EncryptionKey/Setup/Patch/Data/SodiumChachaPatch.php:
```diff
@@ -0,0 +1,129 @@
+<?php
+
```
Strict type check must be enabled for all PHP files.

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+    ) {
+        $this->cipher = $cipher;
+        $this->mode = $mode;
+        // @codingStandardsIgnoreStart
```
With `@codingStandardIgnoreLine` the code will look cleaner, in case you need to ignore just one line.

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+            // @codingStandardsIgnoreEnd
+            if (true === $initVector) {
+                /* Generate a random vector from human-readable characters */
+                $abc = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
```
`$abc` => `allowedCharacters` ?

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+                $abc = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
+                $initVector = '';
+                for ($i = 0; $i < $initVectorSize; $i++) {
+                    $initVector .= $abc[rand(0, strlen($abc) - 1)];
```
`rand()` is not cryptographically secure, please consider replacing it with [random_bytes()](http://php.net/manual/en/function.random-bytes.php)

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+     * @param string $key
+     * @param string $cipher
+     * @param string $mode
+     * @param bool $initVector
```
According to the method implementation, `$initVector` could be `true`, `false` or `string`, which does not match the annotation. Also having mixed types for arguments is complicating the logic and is confusing.

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+            if (true === $initVector) {
+                /* Generate a random vector from human-readable characters */
+                $abc = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
+                $initVector = '';
```
Changing variable type from `bool` to `string` on the fly can be confusing and potentially lead to bugs.

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+                    )
+                );
+            }
+            $this->_initVector = $initVector;
```
`_initVector` is undefined, `initVector` was supposed to be used instead.

paliarush commented on lib/internal/Magento/Framework/Encryption/Adapter/Mcrypt.php:
```diff
@@ -0,0 +1,164 @@
+    public function encrypt($data)
+    {
+        throw new \Exception(
+            (string)new \Magento\Framework\Phrase('Mcrypt cannot be used for encryption. Use Sodium instead')
```
If this is logic exception, no need to wrap the message into `Phrase` object (and as a result add it to the translation dictionary)

pmclain committed:
Remove CC Number Re-encryption

Will be implemented as a CLI tool instead

pmclain committed:
Add strict types

pmclain committed:
Replace autogenerated doc blocks

With actual class and function definitions

pmclain committed:
Add param and return types for new methods

pmclain committed:
Make encrypt and decrypt symetric

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento-engcom/php-7.2-support?pullRequest=135) <br/>All committers have signed the CLA.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento-engcom/php-7.2-support?pullRequest=135) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-engcom/php-7.2-support?pullRequest=135) before we can accept your contribution.<br/>

pmclain committed:
Replace single line codeStandardIgnore with codingStandardIgnoreLine

pmclain committed:
Add Crypt to obsolete class list

pmclain committed:
Change return types of getCrypt and validateKey

`Magento\Framework\Encryption\Encryptor::validateKey` - Now returns `void`
instead of an instance of `\Magento\Framework\Encryption\Crypt`.

`Magento\Framework\Encryption\Encryptor::getCrypt` - Scope was changed from
`protected` to `private` and return value is now an instance of
`Magento\Framework\Encryption\Adapter\EncryptionAdapterInterface`

pmclain committed:
Add CLI Command for Updating CC Number Encryption

Updates the encryption of any values in `sales_order_payment.cc_number_enc`
with a cipher version below the latest. The command executes in batches of 1000

pmclain committed:
Fix property assignment in Mcrypt adapter

buskamuza commented on app/code/Magento/Sales/Console/Command/PaymentUpdateEncryptCommand.php:
```diff
@@ -0,0 +1,95 @@
+class PaymentUpdateEncryptCommand extends Command
+{
+    /** Command name */
+    const NAME = 'encryption:update:payment';
```
Please follow https://devdocs.magento.com/guides/v2.2/extension-dev-guide/cli-cmds/cli-naming-guidelines.html
Maybe something like `encryption:payment-data:update`
cc @piotrekkaminski

piotrekkaminski replied:
encryption:payment-data:update (or reencrypt) looks good to me. Is it the only CLI command? how are stored credentials for eg PayPal updated? What if we import file containing old password hashes?

pmclain replied:
config values using `\Magento\Config\Model\Config\Backend\Encrypted` as the backend model and re-encrypted with `Magento\EncryptionKey\Setup\Patch\Data\SodiumChachaPatch`. The payment data was removed from the patch over concerns about the volume of data. Hashing functions are outside the scope of these changes and have not been altered.

buskamuza commented on app/code/Magento/Sales/Console/Command/PaymentUpdateEncryptCommand.php:
```diff
@@ -0,0 +1,95 @@
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $connection = $this->paymentResource->getConnection();
```
DB operations are too low-level to be in a CLI command (as it should be considered the same as a controller).
Could you please move DB logic to a model?

buskamuza commented on app/code/Magento/EncryptionKey/Setup/Patch/Data/SodiumChachaPatch.php:
```diff
@@ -0,0 +1,117 @@
+                \Magento\Config\Model\Config\Backend\Encrypted::class
+            );
+        } catch (\Exception $e) {
+            // This is thrown during initial application installation
```
It's not allowed to use exceptions for implementing business logic. I propose to fix https://github.com/magento-engcom/php-7.2-support/blob/9f76f7cce2773c6367bda42a0d54dd50efbcee52/app/code/Magento/Config/Model/Config/Structure.php#L284 , so it doesn't break when the index is absent.

buskamuza commented on app/code/Magento/Sales/Console/Command/PaymentUpdateEncryptCommand.php:
```diff
@@ -0,0 +1,95 @@
+            )->limit(1000);
+
+        while ($attributeValues = $connection->fetchPairs($select)) {
+            $connection->beginTransaction();
```
As the command can be run by somebody on a live store, transactions may cause performance issues/locks.
I propose to remove the transaction, and add another condition to the update, so that we guaranteed do not replace data that a user could enter if it happens between we fetch the data and update it. Something like this:
replace `['entity_id = ?' => (int)$valueId]` with `['entity_id = ?' => (int)$valueId, 'and cc_number_enc = ?' => $value]`

pmclain committed:
Return empty array if sections key is empty

The `sections` key may not be set during module installation or upgrade and
would result in an undefined index notice.

pmclain committed:
Refactor Crypt class to use Mcrypt Adapter

Removes code duplication between `Crypt` and `Mcrpyt` classes. `Mcrpyt` is used
for all equivalent `Crypt` methods except `Crypt::encrypt` becuase
`Mcrypt::encrpt` throws an exception preventing it's use.

pmclain committed:
Include Cipher In Sodium classname

Unlike mcrypt, the sodium extension uses a different [en|de]crypt function for
each avaiable cipher. Including the cipher in the class name provides a method
for changing the chosen sodium cipher.

buskamuza committed:
Merge remote-tracking branch 'mainline/2.3-develop' into pmclain-2.3.1/libsodium

pmclain committed:
Unit tests for encryption adapters

pmclain committed:
Integration Test for CC Number Encryption Updater

pmclain committed:
Integration Test for Config Encrypt Update Patch

Test validates the following are true after patch is applied:
 * Patched values are encrypted with latest cipher
 * Patched values match original plain-text value when decrpyted

pmclain committed:
Redistribute integration tests in travis builds

rganin commented on app/code/Magento/EncryptionKey/Setup/Patch/Data/SodiumChachaPatch.php:
```diff
@@ -0,0 +1,112 @@
+    {
+        $currentScope = $this->scope->getCurrentScope();
+
+        $this->scope->setCurrentScope(\Magento\Framework\App\Area::AREA_ADMINHTML);
```
Area is not reset
